elastic/kibana
# [Security Solution][Legacy Actions] - Update legacy action migration to account for more edge cases

## Summary

Updates the legacy actions migration code to account for edge cases we had not initially caught. Thanks to testing from some teammates, they reported seeing the following behavior:

- Rules created pre 7.16 with no actions still create the legacy action sidecar (but not a `siem.notifications` legacy actions alert) which upon migration to 7.16+ was not being deleted
- Rules created pre 7.16 with actions that run on every rule run create the legacy action sidecar(but not a `siem.notifications` legacy actions alert) which upon migration to 7.16+ was not being deleted
- Rules created pre 7.16 with actions that were never enabled until 8.x did not have a `siem.notifications` legacy actions alert type created

Because the legacy migration code relied on checking if a corresponding `siem.notifications` SO existed to kick off the necessary cleanup/migration, the above edge cases  were not being caught.

Working to migrate the legacy actions can definitely scramble your head a bit so I'll try to break it down below and made sure to add plenty of tests (e2e, unit) to check the logic.

| Actions        | Pre migrated rule           | Post migrated rule  |
|:----:|:-----:|:-----:|
| None      | - NO `siem.notifications` SO created <br />- `siem-detection-engine-rule-actions` SO created <br />- Empty actions array on rule params  | -  `siem-detection-engine-rule-actions` DELETED  <br />- `throttle` is `null` <br />- `notifyWhen` is `onActiveAlert` <br />- `actions` is `[]` |
| Rule run      | - NO `siem.notifications` SO created  <br />- `siem-detection-engine-rule-actions` SO created  <br />- action lives on rule params   |   -  `siem-detection-engine-rule-actions` DELETED <br />- actions continue to live on rule params <br />- `throttle` is `null` <br />- `notifyWhen` is `onActiveAlert` |
| Hourly | - `siem.notifications` SO created   <br />- `siem-detection-engine-rule-actions` SO created  <br />- actions array on rule params empty |    -  s`iem-detection-engine-rule-actions` DELETED <br />- `siem.notifications` DELETED<br /> - actions moved to live on rule params  <br />- `throttle` is `1h` <br />- `notifyWhen` is `onThrottleInterval` |
| Daily | - `siem.notifications` SO created   <br />- `siem-detection-engine-rule-actions` SO created  <br />- actions array on rule params empty |    -  `siem-detection-engine-rule-actions` DELETED <br />- `siem.notifications` DELETED<br /> - actions moved to live on rule params  <br />- `throttle` is `1d` <br />- `notifyWhen` is `onThrottleInterval` |
| Weekly | - `siem.notifications` SO created   <br />- `siem-detection-engine-rule-actions` SO created  <br />- actions array on rule params empty |    -  `siem-detection-engine-rule-actions` DELETED <br />- `siem.notifications` DELETED<br /> - actions moved to live on rule params  <br />- `throttle` is `7d` <br />- `notifyWhen` is `onThrottleInterval` |

## Testing

- Check out 7.15 branch locally
- Create 5 enabled rules: with no actions, with an action triggered per each rule execution, with an action triggered every hour, every day, every week
- Use the scripts below in Dev Tools to check that conditions outlined in the above table are true
- Check out the 7.17 branch locally (required to go to 7.17 before jumping to 8.0)
- Upgrade the stack to 7.17 - run ES, then run Kibana
- Check out this feature branch and upgrade the stack to 8.2.0 - run ES, then run Kibana
- Execute the same requests in Dev Tools and verified that results match those outlined in table above under "Pre migrated rule"
- Bulk enabled all the rules
- Execute the same requests in Dev Tools and verified that results match those outlined in table above under "Post migrated rule"

Useful scripts:
```
## Gets the alerting rule type created for the legacy action
GET .kibana/_search
{
  "query": {
    "term": {
      "alert.alertTypeId": "siem.notifications"
    }
  }
}

## Gets the siem sidecar actions
GET .kibana/_search
{
  "query": {
    "term": {
      "type": {
        "value": "siem-detection-engine-rule-actions"
      }
    }
  }
}

## Gets the rules pre 8.0
## For legacy actions with an interval other than on every rule run, the `actions` array is empty and 
## `throttle : null`, and `notifyWhen : onActiveAlert`.
GET .kibana/_search
{
  "query": {
    "term": {
      "alert.alertTypeId": "siem.signals"
    }
  }
}

## Gets the query rules post 8.0
## For NON legacy actions that have successfully migrated actions, the `actions` array is filled out on the rule,
## the `throttle` corresponds to the selected interval and `notifyWhen: onThrottleInterval`
GET .kibana/_search
{
  "query": {
    "term": {
      "alert.alertTypeId": "siem.queryRule"
    }
  }
}
```

## Followup Needed

- [ ] Performance testing

##  Checklist


### For maintainers
